pkg/ring/ring.go

```diff
@@ -363,29 +363,73 @@ func (r *Ring) GetReplicationSetForOperation(op Operation) (ReplicationSet, erro
 		return ReplicationSet{}, ErrEmptyRing
 	}
 
-	// Calculate the number of required ingesters;
-	// ensure we always require at least RF-1 when RF=3.
-	numRequired := len(r.ringDesc.Ingesters)
-	if numRequired < r.cfg.ReplicationFactor {
-		numRequired = r.cfg.ReplicationFactor
-	}
-	maxUnavailable := r.cfg.ReplicationFactor / 2
-	numRequired -= maxUnavailable
-
-	ingesters := make([]IngesterDesc, 0, len(r.ringDesc.Ingesters))
+	// Build the initial replication set, excluding unhealthy instances.
+	instances := make([]IngesterDesc, 0, len(r.ringDesc.Ingesters))
+	zoneFailures := make(map[string]struct{})
 	for _, ingester := range r.ringDesc.Ingesters {
 		if r.IsHealthy(&ingester, op) {
-			ingesters = append(ingesters, ingester)
+			instances = append(instances, ingester)
+		} else {
+			zoneFailures[ingester.Zone] = struct{}{}
 		}
 	}
 
-	if len(ingesters) < numRequired {
-		return ReplicationSet{}, ErrTooManyFailedIngesters
+	// Max errors and max unavailable zones are mutually exclusive. We initialise both
+	// to 0 and then we update them whether zone-awareness is enabled or not.
+	maxErrors := 0
+	maxUnavailableZones := 0
+
+	if r.cfg.ZoneAwarenessEnabled {
+		// Given data is replicated to RF different zones, we can tolerate a number of
+		// RF/2 failing zones. However, we need to protect from the case the ring currently
+		// contains instances in a number of zones < RF.
+		numReplicatedZones := util.Min(len(r.ringZones), r.cfg.ReplicationFactor)
+		minSuccessZones := (numReplicatedZones / 2) + 1
```

Comment:
Added a special case for rf=2 and nr of zones =2/

Reply:
I'm actually not sure anymore if my previous statement is correct. Reason is that during write, we would accept it if we get *single* OK. But during read, we don't know which single zone has received the sample, so we need to ask both.

I'd check with @pracucci to be 100% sure. But now I think my previous comment here was incorrect.

Reply:
Changed

Reply:
What's the current status of this conversation?

`DefaultReplicationStrategy.Filter()` calculate `minSuccess := (replicationFactor / 2) + 1`. The `minSuccess` is the write quorum. If RF=2, then we need to write to both ingesters in order to succeed. Because of this, it's safe to have `maxUnavailableZones=1` when `RF=2`.

I suggest changing the logic as follow (the idea is that :
```
maxUnavailableZones = minSuccessZones - 1
```

Tests should be updated accordingly.

Reply:
> What's the current status of this conversation?

I convinced myself that with RF=2, it is enough to write to single ingester. But that's not a majority 🤦